universal-application-tool-0.0.1/app/services/role/RoleService.java

```diff
@@ -63,14 +63,17 @@ public RoleService(ProgramService programRepository, UserRepository userReposito
             .filter(address -> !Strings.isNullOrEmpty(address))
             .collect(toImmutableSet());
     ImmutableSet.Builder<String> invalidEmailBuilder = ImmutableSet.builder();
-    accountEmails.forEach(
-        email -> {
-          if (sysAdminEmails.contains(email)) {
-            invalidEmailBuilder.add(email);
-          } else {
-            userRepository.addAdministeredProgram(email, program);
-          }
-        });
+    for (String email: accountEmails){
+      if (sysAdminEmails.contains(email)) {
+        invalidEmailBuilder.add(email);
+      } else {
+        Optional<CiviFormError> maybeError = userRepository.addAdministeredProgram(email, program);
+        // If there was an error adding the administered program, return the error. 
+        if (maybeError.isPresent()){
+          return maybeError;
+        }
```

Comment:
suggestion (blocking): The behavior here is a little unpredictable -- if the list of emails has valid emails A and B and invalid email I and the list is in the order AIB then A will get updated correctly but B won't due to the early return triggering for "I". Suggest creating a new error that combines messages as needed if there is more than one error and looping through the entire list regardless of validity.

Reply:
Really good point. Made a fix, let me know if this works.